### PR TITLE
Testing modification for slurm GPU features

### PR DESCRIPTION
**Pull request by rexcsn**
* Added new feature test slurm_gpu in test_slurm.py
* Modified _test_dynamic_dummy_nodes and related functions to check for GPU resources when requested
* Added a note about debugging a specific failures in testing framework

Signed-off-by: Rex <shuningc@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 1]
```diff
@@ -38,19 +40,129 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+
+    assert_no_errors_in_logs(remote_command_executor, ["/var/log/sqswatcher", "/var/log/jobwatcher"])
+
+
+def _gpu_test_cluster_limits(remote_command_executor, max_queue_size, num_gpus):
+    """Test edge cases regarding the number of GPUs."""
+    logging.info("Testing scheduler does not accept jobs when requesting for more GPUs than available")
+
+    # Expect commands below to fail with exit 1
+    result = remote_command_executor.run_remote_command(
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: Extend submit_job to take GPU options.

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 2]
```diff
@@ -38,19 +40,129 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+        "sbatch -N 1 --wrap='sleep 1' --gpus-per-task {0}".format(num_gpus + 1), raise_on_error=False
+    )
+    assert_that(result.stdout).contains(
+        "sbatch: error: Batch job submission failed: Requested node configuration is not available"
+    )
+    result = remote_command_executor.run_remote_command(
+        "sbatch -N 1 --wrap='sleep 1' --gres=gpu:{0}".format(num_gpus + 1), raise_on_error=False
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: same here

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 3]
```diff
@@ -38,19 +40,129 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+        "sbatch -G {0} --gpus-per-node={1} --wrap='sleep 1'".format(num_gpus * max_queue_size, num_gpus)
+    )
+
+
+def _gpu_test_conflicting_options(remote_command_executor, num_gpus):
+    """Test GPU-related conflicting option senerios."""
+    logging.info("Testing scheduler does not accept jobs when requesting job containing conflicting options")
+
+    result = remote_command_executor.run_remote_command(
+        "sbatch -G 1 --cpus-per-gpu 32 -N 1 --cpus-per-task 20 --wrap='sleep 1'", raise_on_error=False
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: Maybe it would be useful to extend `submit_script` to take arbitrary args.

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 4]
```diff
@@ -73,21 +185,33 @@ def _test_dynamic_max_cluster_size(remote_command_executor, region, asg_name):
     time.sleep(200)
     _assert_dummy_nodes(remote_command_executor, new_max_size)
 
+    # Check slurmctld start with no node
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: comments?

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 5]
```diff
@@ -160,24 +284,46 @@ def _retrieve_slurm_compute_nodes_from_config(remote_command_executor):
...
+
+
+@retry(wait_fixed=seconds(20), stop_max_delay=minutes(7))
+def _assert_asg_has_no_node(region, asg_name):
+    assert_asg_desired_capacity(region, asg_name, expected=0)
+
+
+def _assert_dummy_nodes(remote_command_executor, count, slots=4, gpus=0):
     __tracebackhide__ = True
+    if gpus != 0:
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: > 0

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 6]
```diff
@@ -38,19 +40,129 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+    """Test cluster is scaling up correctly and GPU jobs are not aborted on slurmctld restart."""
+    logging.info("Testing cluster scales correctly with GPU jobs")
+    slurm_commands = SlurmCommands(remote_command_executor)
+    # Assert initial conditions
+    _assert_asg_has_no_node(region, asg_name)
+    _assert_no_nodes_in_scheduler(slurm_commands)
+
+    # Should require 3 nodes
+    result = remote_command_executor.run_remote_command(
+        "sbatch --wrap 'sleep 10' -c 10 -n 4 -G {0}".format(num_gpus * 2 + 1)
```
Comment (demartinofra) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: Add tests for these 3 jobs, submitted at the same time: --gpus=3, --gpus-per-task=1 -n 3, --gpus-per-node 1 -N 1.

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 1 above)
Comment (rexcsn) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: I don't think abstraction is very valuable in this case. There are many slurm options that we are only using for GPU tests. We will have to end up either setup a variable in `submit_job` for each of these option or setup an `arbitrary_args` variable. However either way we would ended up almost typing the entire command anyway, and I think that's not as clear as just seeing the command as is.

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 7]
```diff
@@ -160,27 +309,54 @@ def _retrieve_slurm_compute_nodes_from_config(remote_command_executor):
     return [node for node in slurm_nodes.splitlines() if "dummy-compute" not in node]
 
 
-def _retrieve_slurm_nodes_from_config(remote_command_executor):
-    retrieve_nodes_command = "sudo grep 'NodeName' /opt/slurm/etc/slurm_parallelcluster_nodes.conf"
+def _retrieve_slurm_nodes_from_config(remote_command_executor, gres=False):
+    if gres:
+        retrieve_nodes_command = "sudo grep 'NodeName' /opt/slurm/etc/slurm_parallelcluster_gres.conf"
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: I'd use python regex / contains() instead of grep. This is because when it fails, it's much easier to debug, since the error message contains the full output.

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 8]
```diff
@@ -46,11 +48,148 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+    # Assert scaling worked as expected
+    assert_scaling_worked(slurm_commands, region, stack_name, scaledown_idletime, expected_max=4, expected_final=0)
+    # Assert jobs were completed
+    for job_id in job_ids:
+        _assert_job_completed(remote_command_executor, job_id)
+
+
+def _submit_command_and_get_job_id(remote_command_executor, slurm_commands, command):
+    """Submit job and get job ID."""
+    result = remote_command_executor.run_remote_command("{0}".format(command))
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: Just use slurm submit_command() / assert_job_submitted()

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 9]
```diff
@@ -46,11 +48,148 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+        _submit_command_and_get_job_id(remote_command_executor, slurm_commands, "sbatch --wrap 'sleep 10' -c 32 -n 1")
+    )
+    # Nodes/resources available after this job:
+    # [{cpu:31, gpu:0}, {cpu:31, gpu:0}, {cpu:21, gpu:0}, {cpu:0, gpu:2}]
+
+    # Assert scaling worked as expected
+    assert_scaling_worked(slurm_commands, region, stack_name, scaledown_idletime, expected_max=4, expected_final=0)
+    # Assert jobs were completed
+    for job_id in job_ids:
+        _assert_job_completed(remote_command_executor, job_id)
```
Comment (sean-smith) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: use `assert_job_succeeded()` from schedulers_common.py

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 1 above)
Comment (sean-smith) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: I agree with you for some of the commands, for others I'd lean towards using the scheduler functions because it returns the job id, allows you to check the job status easily, and is centralized logic.

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 10]
```diff
@@ -38,19 +40,129 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
...
+    _test_dynamic_dummy_nodes(remote_command_executor, region, cluster.asg, max_queue_size)
 
     assert_no_errors_in_logs(remote_command_executor, ["/var/log/sqswatcher", "/var/log/jobwatcher"])
 
 
+@pytest.mark.regions(["us-east-2"])
+@pytest.mark.instances(["g3.8xlarge"])
+@pytest.mark.schedulers(["slurm"])
+@pytest.mark.usefixtures("os", "instance", "scheduler")
+@pytest.mark.slurm_gpu
```
[the remark this answers is not in the document]
Comment (rexcsn) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: Using this marker so we can run `slurm_gpu` test as a feature by itself. Testing framework only parses feature by filename, functions inside a file are not parsed as features

File: tests/integration-tests/tests/common/utils.py [hunk 11]
```diff
@@ -28,3 +28,17 @@ def fetch_instance_slots(region, instance_type):
             "Could not load instance mapping file from S3 bucket {0}. Failed with exception: {1}".format(bucket_name, e)
         )
         raise
+
+
+def fetch_instance_gpus(region, instance_type):
+    bucket_name = "{0}-aws-parallelcluster".format(region)
+    try:
+        s3 = boto3.resource("s3", region_name=region)
```
[the remark this answers is not in the document]
Comment (rexcsn) [reply tests/integration-tests/tests/common/utils.py]: extended `fetch_instance_slots` to take in arbitrary type of `slots`

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 12]
```diff
@@ -38,19 +40,129 @@ def test_slurm(region, pcluster_config_reader, clusters_factory):
     cluster = clusters_factory(cluster_config)
     remote_command_executor = RemoteCommandExecutor(cluster)
 
-    _test_slurm_version(remote_command_executor)
+    # _test_slurm_version(remote_command_executor)
```
[the remark this answers is not in the document]
Comment (rexcsn) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: uncommented

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 13]
```diff
@@ -160,27 +309,54 @@ def _retrieve_slurm_compute_nodes_from_config(remote_command_executor):
...
+        )
+    else:
+        assert_that(dummy_nodes_config[0]).is_equal_to(
+            "NodeName=dummy-compute[1-{0}] CPUs={1} {2}State=FUTURE".format(count, slots, gpu_entry)
+        )
+        dummy_nodes_count = _retrieve_slurm_dummy_nodes(remote_command_executor)
+        assert_that(dummy_nodes_count).is_equal_to(count)
 
 
 def _get_job_info(remote_command_executor, job_id):
```
[the remark this answers is not in the document]
Comment (rexcsn) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: moved and renamed to `get_job_info`

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 7 above)
Comment (rexcsn) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: Addressed, doing a `cat` file and handling all string logic in python
